### PR TITLE
ARROW-459: [C++] Dictionary IPC support in file and stream formats

### PR DESCRIPTION
**Pull request by wesm**
Also fixes ARROW-565

### CODE REVIEW

**Comment by wesm** [comment]
OK, I've done as much as I think I should in this patch. We'll be forced to resolve any lingering issues when we get to making the integration tests run between C++/Java

File: cpp/src/arrow/ipc/metadata.h [hunk 1]
```diff
@@ -47,9 +52,38 @@ struct MetadataVersion {
...
+class DictionaryMemo {
+ public:
+  DictionaryMemo();
+
+  // Returns KeyError if dictionary not found
+  Status GetDictionary(int64_t id, std::shared_ptr<Array>* dictionary) const;
+
+  int64_t GetId(const std::shared_ptr<Array>& dictionary);
+
+  bool HasDictionary(const std::shared_ptr<Array> dictionary) const;
```
Comment (xhochy) [inline cpp/src/arrow/ipc/metadata.h]: missing a reference here

Comment (wesm) [inline cpp/src/arrow/ipc/metadata.h]: done

**Comment by wesm** [comment]
I don't know what is going on with our Travis CI build queue but I opened: https://issues.apache.org/jira/browse/INFRA-13570

**Comment by wesm** [comment]
seems to be an ASF build queue problem. the build for my latest revision is running here: https://travis-ci.org/wesm/arrow/builds/204620814

**Comment by wesm** [comment]
@xhochy for large code reviews in the future, reminder that we can always do reviews on Cloudera's Gerrit instance, where we have a project set up: https://gerrit.cloudera.org/#/admin/projects/. I've been lobbying for an ASF-administered Gerrit (so I can bug them instead of Cloudera folks when I need some admin stuff done) but that may be useful in the meantime

**Comment by xhochy** [comment]
I only went briefly over this. Will have a more close look tomorrow, then I can decide if we need gerrit or not.

**Comment by wesm** [comment]
Rebased (to pick up integration test fix) and fixed valgrind error

**Comment by wesm** [comment]
passing build in my fork: https://travis-ci.org/wesm/arrow/builds/204779749. not sure why Travis is having such a hard time

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**